File: 123-best-time-to-buy-and-sell-stock-iii/best-time-to-buy-and-sell-stock-iii.java
class Solution {
    public int maxProfit(int[] prices) {
        int firstBuy=Integer.MIN_VALUE;
        int firstSell=0;
        int secondBuy=Integer.MIN_VALUE;
        int secondSell=0;
        for(int i:prices){
            firstBuy=Math.max(firstBuy,-i);
            firstSell=Math.max(firstSell,firstBuy+i);
            secondBuy=Math.max(secondBuy,firstSell-i);
            secondSell=Math.max(secondSell,secondBuy+i);
        }
        return secondSell;
    }
}

File: 123-best-time-to-buy-and-sell-stock-iii/best-time-to-buy-and-sell-stock-iii.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int buy1=INT_MIN;
        int sell1=0;
        int buy2=INT_MIN;
        int sell2=0;
        for(int i:prices){
            buy1=max(buy1,-i);
            sell1=max(sell1,buy1+i);
            buy2=max(buy2,sell1-i);
            sell2=max(sell2,buy2+i);
        }
        return sell2;
    }
};